### PR TITLE
[geometry] Update mesh_indices.h to not need a generator file

### PR DESCRIPTION
**Pull request by jmaiolini**
Templated class version of mesh_index.h (issue #4633).

Comments:
*  Kept conversions outside the class (not templated since they are type specific)
*  Kept istream definition outside the templated class by defining it at the beginning (verbose)
* Changed class ns as requested so I changed also the name for the doc generation. I don't know if it has to remain in geometry \ingroup.

### CODE REVIEW

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 1]
```diff
@@ -38,314 +38,54 @@
...
 ////////////////////////////////////////////////////////////////////////////////
 
 namespace pcl {
-namespace geometry {
+namespace detail {
 /** \brief Index used to access elements in the half-edge mesh. It is basically just a
  * wrapper around an integer with a few added methods.
  * \author Martin Saelzle
- * \ingroup geometry
+ * \ingroup detail
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:56]: You can just repeat the doc-string near the declarations (L189)

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 2]
```diff
@@ -38,314 +38,54 @@
...
- * wrapper around an integer with a few added methods.
- * \author Martin Saelzle
- * \ingroup geometry
- */
-class HalfEdgeIndex
+template <class MeshIndexT>
+class MeshIndex
 : boost::totally_ordered<
-      pcl::geometry::HalfEdgeIndex // < > <= >= == !=
+      pcl::detail::MeshIndex<MeshIndexT> // < > <= >= == !=
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:68]: Could you see if we can skip `pcl::detail::` (and move the comma to the previous line)

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 3]
```diff
@@ -426,167 +166,43 @@ class EdgeIndex
...
   return (is >> index.index_);
 }
 
-} // End namespace geometry
-} // End namespace pcl
-
-////////////////////////////////////////////////////////////////////////////////
-// FaceIndex
-////////////////////////////////////////////////////////////////////////////////
+using VertexIndex = MeshIndex<struct VertexIndexT>;
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:188]: PCL uses `PrefixT` in a lot of places like `PointT` to mean it's a template type argument for a point. To prevent confusion, can we rename to `VertexIndexTag` or similar?

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 4]
```diff
@@ -597,12 +213,12 @@ toEdgeIndex(const HalfEdgeIndex& index)
...
 toHalfEdgeIndex(const EdgeIndex& index, const bool get_first = true)
 {
   return (index.isValid()
               ? HalfEdgeIndex(index.get() * 2 + static_cast<int>(!get_first))
               : HalfEdgeIndex());
 }
-} // End namespace geometry
-} // End namespace pcl
+} // End namespace detail
+} // End namespace pcl
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:224]: new lines needed at the end of file

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
🥳 Woo!! 35 lines added to replace 419 lines (and counting) 😄 

You need to run clang-format on the file (using `make format`) or just apply the patch that is generated on the format CI

Please also remove the python file in this PR itself (using `git rm <path-to-file>`)

**Comment by jmaiolini** [comment]
Coding sorceries 😂 😂 
Will do all the above as soon as possible. Sorry didn't know about the clang-format requirement (however it is not present inside CONTRIBUTING.md or I missed it).

File: geometry/include/pcl/geometry/mesh_indices.h (hunk 1 above)
Comment (jmaiolini) [reply geometry/include/pcl/geometry/mesh_indices.h:56]: Regarding this modification, what should I do? Do you want me to put the `\ingroup detail` for each class method?

Comment (kunaltyagi) [reply geometry/include/pcl/geometry/mesh_indices.h:56]: in group geometry, just like usual. Basically, no change in the generated documentation

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 5]
```diff
@@ -425,184 +163,58 @@ class EdgeIndex
...
-/** \brief istream operator. */
-inline std::istream&
-operator>>(std::istream& is, pcl::geometry::FaceIndex& index)
-{
-  return (is >> index.index_);
-}
+using VertexIndex = MeshIndex<struct VertexIndexTag>;
+using HalfEdgeIndex = MeshIndex<struct HalfEdgeIndexTag>;
+using EdgeIndex = MeshIndex<struct EdgeIndexTag>;
+using FaceIndex = MeshIndex<struct FaceIndexTag>;
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:189]: These should be in pcl::geometry

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 6]
```diff
@@ -38,314 +38,52 @@
...
-namespace geometry {
-/** \brief Index used to access elements in the half-edge mesh. It is basically just a
- * wrapper around an integer with a few added methods.
- * \author Martin Saelzle
- * \ingroup geometry
+namespace detail {
+/** \brief Index used to access elements in the half-edge mesh. It is basically
+ * just a wrapper around an integer with a few added methods. \author Martin
+ * Saelzle \ingroup geometry
  */
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:56]: Move these to L186-L189

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 7]
```diff
@@ -425,184 +163,58 @@ class EdgeIndex
...
+} // End namespace detail
 } // End namespace pcl
 
 ////////////////////////////////////////////////////////////////////////////////
 // Conversions
 ////////////////////////////////////////////////////////////////////////////////
 
 namespace pcl {
-namespace geometry {
+namespace detail {
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:199]: Don't change the namespace of old things

**Review by kunaltyagi — CHANGES_REQUESTED** [review]

**Comment by jmaiolini** [comment]
Yep, now it makes much more sense what you wanted to achieve 😄 
Thank you for your patience!

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 8]
```diff
@@ -183,20 +180,29 @@ operator>>(std::istream& is, MeshIndex<MeshIndexTag>& index)
...
-using EdgeIndex = MeshIndex<struct EdgeIndexTag>;
-using FaceIndex = MeshIndex<struct FaceIndexTag>;
-
 } // End namespace detail
 } // End namespace pcl
 
+namespace pcl {
+namespace geometry {
+/** \brief Index used to access elements in the half-edge mesh. It is basically
+ * just a wrapper around an integer with a few added methods. \author Martin
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:189]: Could you just  copy paste the doxygen from the original version here and before the other `using` statements?

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 9]
```diff
@@ -425,156 +160,51 @@ class EdgeIndex
...
-
-/** \brief istream operator. */
-inline std::istream&
-operator>>(std::istream& is, pcl::geometry::FaceIndex& index)
-{
-  return (is >> index.index_);
-}
+using VertexIndex = pcl::detail::MeshIndex<struct VertexIndexTag>;
+/** \brief Index used to access elements in the half-edge mesh. It is basically
+ * just a wrapper around an integer with a few added methods. \author Martin
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:194]: If you notice, the original comment has \author and \ingroup on new lines (which is required for doxygen). Could you please make the necessary changes?

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 10]
```diff
@@ -38,314 +38,49 @@
...
-inline std::istream&
-operator>>(std::istream& is, pcl::geometry::VertexIndex& index)
-{
-  return (is >> index.index_);
-}
+namespace detail {
 
-} // End namespace geometry
-} // End namespace pcl
+template <class MeshIndexTag>
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:54]: Since this is a template argument, please call it `IndexTagT` (read as index-tag-type)

File: geometry/include/pcl/geometry/mesh_indices.h (hunk 9 above)
Comment (jmaiolini) [reply geometry/include/pcl/geometry/mesh_indices.h]: You are right. 
I've checked and it is modified when the clang format is applied on the file. Hope now it is ok.

**Review by kunaltyagi — APPROVED** [review]

File: geometry/include/pcl/geometry/mesh_indices.h (hunk 9 above)
Comment (kunaltyagi) [reply geometry/include/pcl/geometry/mesh_indices.h]: The issue is with windows CI not your code

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 11]
```diff
@@ -38,48 +38,49 @@
...
- */
-class VertexIndex
+namespace detail {
+
+template <class IndexTagT>
+class MeshIndex;
+
+template <class IndexTagT>
+std::istream&
+operator>>(std::istream& is, MeshIndex<IndexTagT>&);
```
Comment (mvieth) [inline geometry/include/pcl/geometry/mesh_indices.h:59]: Why are these lines needed? They weren't necessary before, and the class/operator are defined immediately afterwards.

**Review by mvieth — COMMENTED** [review]
Could you rebase these commits onto the current master branch? Then the Windows CI should run successfully.

File: geometry/include/pcl/geometry/mesh_indices.h (hunk 11 above)
Comment (jmaiolini) [reply geometry/include/pcl/geometry/mesh_indices.h:59]: The issue was due to the friend operator overload with a templated argument. The solution to pre-declare the templated operator prevents to have all template instantiations of the class to be friend with all the templated instantiations of the operator ([Johannes Schaub answer](https://stackoverflow.com/questions/1297609/overloading-friend-operator-for-template-class)). A viable alternative is to define it directly into the class.
However, I decided to stick with this solution to keep the structure of the operators from the previous classes, for example I did not declared operator << to be friend, thus for it this issue is not present.

Tell me if you want me to change it, then I will rebase asap!

Comment (mvieth) [reply geometry/include/pcl/geometry/mesh_indices.h:59]: Thanks for the explanation! Then please just rebase onto master to make the CI pass.

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 12]
```diff
@@ -159,422 +160,55 @@ class VertexIndex
   /** \brief Stored index. */
   int index_;
 
-  friend std::istream&
-  operator>>(std::istream& is, pcl::geometry::VertexIndex& index);
+  friend std::istream& operator>>
+      <IndexTagT>(std::istream& is, MeshIndex<IndexTagT>& index);
```
Comment (kunaltyagi) [inline geometry/include/pcl/geometry/mesh_indices.h:164]: From the linked answer, do you think this is a viable alternate?
```suggestion
      <>(std::istream& is, MeshIndex<IndexTagT>& index);
```

Comment (jmaiolini) [reply geometry/include/pcl/geometry/mesh_indices.h:164]: If I remember correctly yes, this works! I can try it and change it on the fly if you want.

File: geometry/include/pcl/geometry/mesh_indices.h (hunk 11 above)
Comment (jmaiolini) [reply geometry/include/pcl/geometry/mesh_indices.h:59]: You are welcome! 
Since I am here, when I rebase to this branch, it says that my branches diverged (which is normal). It is ok to just force push those commits as I would do normally?
I don't know how rebase affects an open PR so just to be sure to do not mess it 😆

Comment (mvieth) [reply geometry/include/pcl/geometry/mesh_indices.h:59]: As dangerous as a force push can be in some other situations, here it is ok. The open PR will be fine :smile:

Comment (kunaltyagi) [reply geometry/include/pcl/geometry/mesh_indices.h:59]: Usually, don't force-push if the PR is in the middle of a review (because GitHub loses count of which files have been marked as read on a force-push)

Once the review is complete and everyone is happy (or if the review hasn't started), you can force-push safely (unless you are collaborating with someone on that branch)

I have a git alias that helps me force-push safely during collaborations:
```
[alias]
  force-push = push --force-with-lease
```

Comment (jmaiolini) [reply geometry/include/pcl/geometry/mesh_indices.h:59]: Rebased smoothly, thank you guys!

**Review by mvieth — APPROVED** [review]
